src/ray/core_worker/store_provider/memory_store/memory_store.cc

```diff
@@ -0,0 +1,206 @@
+#include <condition_variable>
+#include "ray/common/ray_config.h"
+#include "ray/core_worker/context.h"
+#include "ray/core_worker/core_worker.h"
+#include "ray/core_worker/object_interface.h"
+#include "ray/core_worker/store_provider/memory_store_provider.h"
+
+namespace ray {
+
+/// A class that represents a `Get` reuquest.
+class GetRequest {
+ public:
+  GetRequest(const std::vector<ObjectID> &object_ids, bool remove_after_get);
+
+  const std::vector<ObjectID> &ObjectIds() const;
+
+  /// Wait until all requested objects are available, or timeout happens.
+  ///
+  /// \param timeout_ms The maximum time in milliseconds to wait for.
+  /// \return Whether all requested objects are available.
+  bool Wait(int64_t timeout_ms);
+  /// Set the object content for the specific object id.
+  void Set(const ObjectID &object_id, std::shared_ptr<RayObject> buffer);
+  /// Get the object content for the specific object id.
+  std::shared_ptr<RayObject> Get(const ObjectID &object_id) const;
+  /// Whether this is a `get` request.
+  bool ShouldRemoveObjects() const;
+
+ private:
+  /// Wait until all requested objects are available.
+  void Wait();
+
+  /// The object IDs involved in this request. This is used in the reply.
+  const std::vector<ObjectID> object_ids_;
```

Comment:
What happens if object IDs are duplicated in here? Should be documented if that's valid or not.

Reply:
Thanks! this has been fixed to support duplicates.